src/pyscheduling/core/listeners.py

```diff
@@ -0,0 +1,198 @@
+from typing import TextIO
+
+from plotly.offline import init_notebook_mode
+
+import plotly.graph_objects as go
+
+from pyscheduling.Problem import SolveResult, SolveStatus
+
+
+@dataclass
+class BaseListener(ABC):
+
+    solve_result : SolveResult = field(repr=False, init=False) 
+    _start_time: int = field(init=False, repr=False)
+    _end_time: int = field(init=False, repr=False)
+    _total_time: int = field(init=False, repr=False)
+    _nb_sol: int = 0
+    
+    def check_best_sol(self, solution, time_found):
+        
+        if self.solve_result.best_solution is None or solution.objective_value <= self.solve_result.best_solution.objective_value:
+            self.solve_result.best_solution = solution
+
+            if self.solve_result.time_to_best < time_found:
+                self.solve_result.time_to_best = time_found
+
+            return True
+
+        return False
+
+    def on_start(self, solve_result):
+        self.solve_result = solve_result
+        self._start_time = perf_counter()
+    
+    def on_complete(self):
+        self._end_time = perf_counter()
+        self._total_time = self._end_time - self._start_time
+        self.solve_result.runtime = self._total_time
+        self._search_speed = self._nb_sol / self.solve_result.runtime
+        if self._nb_sol > 0:
+            self.solve_result.solve_status = SolveStatus.FEASIBLE
+
+    @abstractmethod
+    def on_solution_found(self, new_solution, time_found, found_new_best):
+        pass
+
+@dataclass
+class FileListener(BaseListener):
+
+    file_path: Path = field(default=Path("solver_log.txt"))
+    _file: TextIO = field(init=False, repr=False)
+
+    def on_start(self, solve_result):
+        super().on_start(solve_result)
+
+        try:
+            self._file = open(self.file_path, "w+")
+            self._file.write(
+                f"{'Info':>15}  |"+
+                f"{'Iteration':>10}  |"+
+                f"{'Time (s)':>10}  |"+
+                f"{'Objective':>10}  |"+
+                f"{'Best':>10}" + "\n\n"
+            )
+        except IOError:
+            print(f"ERROR: cannot open the logging file {self.file_path}")
+    
+    def on_complete(self):
+        super().on_complete()
+
+        self._file.write("\n" + '-' * (80) + "\n\n")
+        
+        self._file.write(f"{'Search completed':<20} : {self.solve_result.nb_solutions} solution(s) found \n"+
+                         f"{'Status code':<20} : {self.solve_result.solve_status}\n\n")
+
+        self._file.write(f"{'Best objective':<20} : {self.solve_result.best_solution.objective_value}\n"+
+                         f"{'Found at time':<20} : {self.solve_result.time_to_best:.2f}\n\n")
+        
+        self._file.write(f"{'Total time':<20} : {self.solve_result.runtime:.2f}\n"+
+                         f"{'Search speed':<20} : {self._search_speed:.2f} sol. / s \n\n")
+        
+        self._file.write("\n" + '-' * (80) + "\n\n")
+
+        if not self._file is None:
```

Comment:
minor: preference to use 
```python
is not None
```

Reply:
Can the file be None at this point? Since the previous statement writes to it. Maybe go straight into the try block